### PR TITLE
Support Rails 7.1, drop Ruby 2.x and Rails < 6.1

### PR DESCRIPTION
**Pull request by eraffel-MDSol**

### CODE REVIEW

**Review by JakubF — CHANGES_REQUESTED** [review]

File: lib/apartment/active_record/connection_handling.rb
Comment (eraffel-MDSol) [reply lib/apartment/active_record/connection_handling.rb:8]: Changing this method would be out of the scope of my PR, which is just to get Rails 7.1 supported. The only reason there's changes in this file is I got rid of the `if` condition which affected the indentation of the rest of the file

File: lib/apartment/active_record/schema_migration.rb
Comment (JakubF) [inline lib/apartment/active_record/schema_migration.rb:7]: I believe this monkey-patch needs to be removed as it breaks rails 7.1 support
This is basically an implementation of `table_exists?` from rails 5.2 and has worked for rails < 7.1 because `SchemaMigration` was inheriting from `ActiveRecord::Base` that was implementing `connection` method, after your PR, this method will start raising error "undefined local variable or method `connection' for ActiveRecord::SchemaMigration:Class"
Current implementation of table_exists is here:
```ruby
    def table_exists?
      @pool.with_connection do |connection|
        connection.data_source_exists?(table_name)
      end
    end
```
it's almost identical - the main difference is that it uses `data_source_exists?` on connection while this monkey patch is forcing it to use `table_exists?`
The difference between these 2 methods is here:
```ruby
      # Checks to see if the data source +name+ exists on the database.
      #
      #   data_source_exists?(:ebooks)
      #
      def data_source_exists?(name)
        query_values(data_source_sql(name), "SCHEMA").any? if name.present?
      rescue NotImplementedError
        data_sources.include?(name.to_s)
      end

      # Checks to see if the table +table_name+ exists on the database.
      #
      #   table_exists?(:developers)
      #
      def table_exists?(table_name)
        query_values(data_source_sql(table_name, type: "BASE TABLE"), "SCHEMA").any? if table_name.present?
      rescue NotImplementedError
        tables.include?(table_name.to_s)
      end
```
so basically just `data_source_sql(name)` VS `data_source_sql(table_name, type: "BASE TABLE")`, the difference that this `type:` part does, is that on table_exists it only looks up partitioned tables and relational tables, and on data_source_exists it also looks up views and materialised views, so this difference is irrelevant for schema migrations.

I believe this monkey-patch needs to be removed, this will 100% cause errors on rails 7.1

**Review by JakubF — CHANGES_REQUESTED** [review]

File: lib/apartment/active_record/schema_migration.rb
Comment (eraffel-MDSol) [reply lib/apartment/active_record/schema_migration.rb:7]: Do you have a failing test I can add then?

Comment (eraffel-MDSol) [reply lib/apartment/active_record/schema_migration.rb:7]: Do you have a failing test I can add then?

Comment (JakubF) [reply lib/apartment/active_record/schema_migration.rb:7]: Calling `ActiveRecord::SchemaMigration.table_exists?` should be enough,
In rails 7.1 it should raise `undefined local variable or method 'connection' for ActiveRecord::SchemaMigration:Class (NameError)`
In rails 7.0 it should be perfectly fine and return true

Comment (eraffel-MDSol) [reply lib/apartment/active_record/schema_migration.rb:7]: It looks like this was added because `table_exists?` was removed from SchemaMigration in rails 6.1, but it looks like it was added back in 7.0. Also, you're looking at `main` in Rails. you need to look at the 7.1 branch. As far as I can tell, this monkey patch is fine, as it is identical to the method in 7.0 and 7.1, and fixes the issue in 6.1. You're right that when 7.2 comes out that this will need to be changed again, but it seems fine now

**Comment by Amnesthesia** [comment]
We had issues with Rails 7.1 and the postgresql adapter, causing a lot of very heavy internal queries, which we reported to rails/rails but which was closed as an Apartment issue (https://github.com/rails/rails/issues/49976)

Curious, does this PR resolve this?

**Comment by brosintoski** [comment]
@JakubF any idea of when this can get merged?  This is currently a blocker for upgrading my app to 7.1.3, and really would like to avoid forking.

**Comment by Amnesthesia** [comment]
@riggleg maybe?

**Comment by eraffel-MDSol** [comment]
@Amnesthesia I haven't had a chance to look at the issue you linked to yet, just FYI

**Comment by Amnesthesia** [comment]
> @Amnesthesia I haven't had a chance to look at the issue you linked to yet, just FYI

Alright, thanks for the update!

**Comment by Amnesthesia** [comment]
Any progress on this @riggleg @JakubF or any other maintainer here?

**Comment by mnovelo** [comment]
@Amnesthesia Hi, I'm a new maintainer!

Taking a look at the [issue you reported to Rails](https://github.com/rails/rails/issues/49976), my understanding is that Rails 7.1 adds the ability to set the isolation level using Fibers to be Fiber-safe, whereas before, the focus was on Threads. [Threads are still the default for Rails 7.1.](https://github.com/rails/rails/pull/44219/files) 

I don't yet see why you experienced a behavior change when upgrading to Rails 7.1 or why the solution that worked for you was changing to using Fibers instead of Threads. As far as I know, the Apartment gem has not yet done anything to use Fibers over Threads. This may well be outside of the scope of this PR.

File: .circleci/config.yml
Comment (mnovelo) [inline .circleci/config.yml:7]:
```suggestion
      - image: cimg/postgres:12.18
```

File: .circleci/config.yml
Comment (mnovelo) [inline .circleci/config.yml:61]:
```suggestion
              ruby_version: ["ruby:3.1.4", "ruby:3.2.3"]
```

**Comment by mnovelo** [comment]
@eraffel-MDSol from PR #236 it looks like there was an API change to ActiveSupport::LobSubscriber that we should also account for in lib/apartment/log_subscriber.rb for Rails 7.1 

In < 7.1, https://github.com/rails/rails/blob/506462ab13755d9f024e1ddbfc8c58d73e7a1bce/activesupport/lib/active_support/log_subscriber.rb#L139

In 7.1, https://github.com/rails/rails/blob/6f0d1ad14b92b9f5906e44740fce8b4f1c7075dc/activesupport/lib/active_support/log_subscriber.rb#L175

**Comment by mnovelo** [comment]
@Amnesthesia I wonder if what you were experiencing might be related to [this issue](https://github.com/rails-on-services/apartment/issues/181) and have been aggravated by [this change](https://github.com/rails/rails/pull/44219/files#diff-258986eb56946168eedb2cbfd0c74bb7ec73ef1c319bdf1d8b8216d205df1219) in Rails 7.1. [This fix](https://github.com/rails-on-services/apartment/pull/182/files) was proposed, but I wonder if it'd be better to have [something like what Rails is doing for ActiveSupport::IsolatedExecutionState](https://github.com/rails/rails/pull/44219/files#diff-258986eb56946168eedb2cbfd0c74bb7ec73ef1c319bdf1d8b8216d205df1219) is better that allows use of Fibers or Threads.

This could be related to the Rails 7.1 upgrade and could warrant including (unless the change is too big), but it'd be helpful if you could also test in your environment

**Comment by brosintoski** [comment]
@mnovelo is this pr not stable enough to merge?  Currently there is No 7.1 support and it appears all critical issues have been addressed.  

Some co-workers of mine are also looking to open a PR for Trilogy db adapter and waiting on this to merge before doing so.

**Comment by mnovelo** [comment]
@brosintoski you're right. I just realized the logger API change is handled in Rails 7.1 with a warning, which we can address in another PR. 

The issue with threads vs fibers requires more investigation, so I'll merge this in for now to unblock other PRs

**Comment by brosintoski** [comment]
@mnovelo @eraffel-MDSol thanks!!!

**Comment by Amnesthesia** [comment]
> @Amnesthesia I wonder if what you were experiencing might be related to [this issue](https://github.com/rails-on-services/apartment/issues/181) and have been aggravated by [this change](https://github.com/rails/rails/pull/44219/files#diff-258986eb56946168eedb2cbfd0c74bb7ec73ef1c319bdf1d8b8216d205df1219) in Rails 7.1. [This fix](https://github.com/rails-on-services/apartment/pull/182/files) was proposed, but I wonder if it'd be better to have [something like what Rails is doing for ActiveSupport::IsolatedExecutionState](https://github.com/rails/rails/pull/44219/files#diff-258986eb56946168eedb2cbfd0c74bb7ec73ef1c319bdf1d8b8216d205df1219) is better that allows use of Fibers or Threads.
> 
> This could be related to the Rails 7.1 upgrade and could warrant including (unless the change is too big), but it'd be helpful if you could also test in your environment

It sounds very likely that it would be related to this indeed. We're in the process of getting rid of Apartment instead because of the amount of headache it causes us, so I'm not sure how much time I'll be able to spend on testing this

**Comment by brosintoski** [comment]
@Amnesthesia out of curiosity what are you replacing Apartment with?

**Comment by eraffel-MDSol** [comment]
seems like this was merged before I had a chance to read comments from the last few days

**Comment by Amnesthesia** [comment]
> @Amnesthesia out of curiosity what are you replacing Apartment with?

Looking at replacing it with `acts_as_tenant`, or custom column based multitenancy

**Comment by mnovelo** [comment]
@Amnesthesia let us know how it goes! Because of how our security documentation is written, column-based multitenancy isn't an option for us, but having an "escape hatch" would be nice to document since I've seen it come up in a few issues

**Comment by Amnesthesia** [comment]
@mnovelo We have still not been able to move away from Apartment. Just wondering, was the issue I brought up fixed upstream or has anybody else had issues with it?

**Comment by mnovelo** [comment]
@Amnesthesia to my knowledge, no one else has struggled with Fibers. Better Fiber support will be included in the next release which I hope to complete soon

**Comment by otagi** [comment]
We’re stuck at Rails 7.0 as well due to Apartment. Our spec suite fails at `activesupport-7.1.4.1/lib/active_support/concurrency/load_interlock_aware_monitor.rb:55:in lock` when calling Apartment’s `switch!`.

We’re considering a switch to `acts_as_tenant` because of that, although the work required is non-trivial.

**Comment by Amnesthesia** [comment]
> We’re stuck at Rails 7.0 as well due to Apartment. Our spec suite fails at `activesupport-7.1.4.1/lib/active_support/concurrency/load_interlock_aware_monitor.rb:55:in lock` when calling Apartment’s `switch!`.
> 
> We’re considering a switch to `acts_as_tenant` because of that, although the work required is non-trivial.

What does that spec look like?

**Pull request merged**
